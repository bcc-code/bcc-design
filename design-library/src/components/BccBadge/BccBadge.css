@layer components {
  .bcc-badge {
    @apply inline-flex items-center justify-center gap-2.5 rounded-md font-normal leading-none tracking-wide;
  }

  /* Size */
  .bcc-badge-xs {
    @apply text-label-xs px-2 py-0.5;
  }
  .bcc-badge-sm {
    @apply text-label-sm leading-4 px-3 py-1;
  }
  .bcc-badge-md {
    @apply text-label-sm leading-4 px-4 py-2;
  }

  .bcc-badge-xs.bordered {
    @apply border py-[1px];
  }
  .bcc-badge-sm.bordered {
    @apply border py-[3px];
  }
  .bcc-badge-md.bordered {
    @apply border py-[7px];
  }

  /* Variants light */
  .bcc-badge-neutral.light {
    @apply bg-neutral-100 text-neutral-900 border-neutral-200;
  }
  .bcc-badge-danger.light {
    @apply bg-red-100 text-red-700 border-red-300;
  }
  .bcc-badge-warning.light {
    @apply bg-yellow-100 text-yellow-800 border-yellow-400;
  }
  .bcc-badge-success.light {
    @apply bg-green-100 text-green-800 border-green-300;
  }
  .bcc-badge-info.light {
    @apply bg-blue-100 text-blue-800 border-blue-300;
  }
  .bcc-badge-muddy-waters.light {
    @apply bg-muddy-waters-100 text-muddy-waters-600 border-muddy-waters-200;
  }
  .bcc-badge-mongoose.light {
    @apply bg-mongoose-100 text-mongoose-700 border-mongoose-200;
  }
  .bcc-badge-brand.light {
    @apply bg-brand-100 text-brand-700 border-brand-200;
  }

  /* Variants dark */
  .bcc-badge-neutral.dark {
    @apply bg-neutral-500 text-neutral-50 border-neutral-400;
  }
  .bcc-badge-danger.dark {
    @apply bg-red-600 text-red-50 border-red-500;
  }
  .bcc-badge-warning.dark {
    @apply bg-yellow-600 text-yellow-50 border-yellow-500;
  }
  .bcc-badge-success.dark {
    @apply bg-green-600 text-green-50 border-green-500;
  }
  .bcc-badge-info.dark {
    @apply bg-blue-500 text-blue-50 border-blue-400;
  }
  .bcc-badge-muddy-waters.dark {
    @apply bg-muddy-waters-400 text-muddy-waters-50 border-muddy-waters-300;
  }
  .bcc-badge-mongoose.dark {
    @apply bg-mongoose-400 text-mongoose-50 border-mongoose-300;
  }
  .bcc-badge-brand.dark {
    @apply bg-brand-400 text-brand-50 border-brand-300;
  }

  /* Icon */
  .bcc-badge-icon {
    @apply h-3 w-3;
  }
  .bcc-badge-sm .bcc-badge-icon {
    @apply h-4 w-4;
  }
  .bcc-badge-md .bcc-badge-icon {
    @apply h-4 w-4;
  }
}
