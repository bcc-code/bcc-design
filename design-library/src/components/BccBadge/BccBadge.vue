<script setup lang="ts">
import type { VueComponent } from "@/types";

withDefaults(
  defineProps<{
    icon?: VueComponent;
    iconRight?: boolean;
    size?: "xs" | "sm" | "md";
    contrast?: "light" | "dark";
    context?:
      | "neutral"
      | "danger"
      | "warning"
      | "success"
      | "info"
      | "muddy-waters"
      | "mongoose"
      | "brand";
  }>(),
  {
    iconRight: false,
    contrast: "light",
    size: "xs",
    context: "neutral",
  }
);
</script>

<template>
  <div class="bcc-badge" :class="`bcc-badge-${context}-${contrast} bcc-badge-${size}`">
    <component
      v-if="icon"
      :is="icon"
      class="bcc-badge-icon"
      :class="[iconRight ? 'order-3' : 'order-1']"
    />
    <span class="order-2"><slot></slot></span>
  </div>
</template>
