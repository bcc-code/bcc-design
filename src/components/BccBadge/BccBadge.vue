<script setup lang="ts">
import type { Component } from "vue";
import { sizeVariants, variants } from "../../composables/variants";

withDefaults(
  defineProps<{
    icon?: string | Component | Function;
    iconRight?: boolean;
    size?: keyof typeof sizeVariants;
    variant?: keyof typeof variants;
  }>(),
  {
    iconRight: false,
    size: "xs",
    variant: "neutral",
  }
);
</script>

<template>
  <div
    class="inline-flex items-center justify-center rounded-md border py-[.2em] px-[1em] font-medium leading-none tracking-wide"
    :class="[variants[variant], sizeVariants[size] ?? size]"
  >
    <component
      v-if="icon"
      :is="icon"
      class="h-[1.2em] w-[1.2em] shrink-0"
      :class="[iconRight ? 'order-3 ml-[.4em]' : 'order-1 mr-[.4em]']"
    />
    <span class="order-2"><slot></slot></span>
  </div>
</template>
