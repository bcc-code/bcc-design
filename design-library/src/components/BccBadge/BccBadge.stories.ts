import BccBadge from "./BccBadge.vue";
import { CheckCircleIcon } from "@bcc-code/icons-vue";

import type { Meta, StoryFn } from "@storybook/vue3";

/**
 * A badge component to be used for informational text, like "new" or "recommended". See the `BccPin` component for a differently styled version for numbers.
 */
export default {
  title: "Common/BccBadge",
  component: BccBadge,
  argTypes: {
    context: {
      description: "Which context color to use for the styling of the badge",
      options: [
        "neutral",
        "danger",
        "warning",
        "success",
        "info",
        "muddy-waters",
        "mongoose",
        "brand",
      ],
      control: { type: "radio" },
    },
    size: {
      description: "Size of the badge",
      options: ["xs", "sm", "md"],
      control: { type: "radio" },
    },
    contrast: {
      description: "contrast",
      options: ["light", "dark"],
      control: { type: "radio" },
    },
    iconRight: {
      description: "Pull icon to right side of the contents",
      control: { type: "boolean" },
    },
    slotDefault: {
      name: "default slot",
      description: "The badge text",
    },
  },
} as Meta<typeof BccBadge>;

const Template: StoryFn<typeof BccBadge> = (args) => ({
  components: { BccBadge },
  setup() {
    return { args, CheckCircleIcon };
  },
  template: `
    <BccBadge v-bind="args" :icon="CheckCircleIcon">
      <template #default>
        {{ args.slotDefault }}
      </template>
    </BccBadge>
  `,
});

export const Example = Template.bind({});
Example.args = {
  context: "success",
  size: "sm",
  iconRight: false,
  slotDefault: "Example Badge",
};
Example.parameters = {
  docs: {
    source: {
      language: "html",
      code: `
<BccBadge context="success" size="sm" :icon="checkCircleIcon">
  Example Badge
</BccBadge>
    `,
    },
  },
};

/**
 * Set the `context` prop to control the color of the badge
 */
export const Context: StoryFn<typeof BccBadge> = () => ({
  components: { BccBadge },
  template: `
    <div class="flex items-start space-x-2">
      <BccBadge context="neutral">neutral</BccBadge>
      <BccBadge context="danger">danger</BccBadge>
      <BccBadge context="warning">warning</BccBadge>
      <BccBadge context="success">success</BccBadge>
      <BccBadge context="info">info</BccBadge>
      <BccBadge context="muddy-waters">muddy-waters</BccBadge>
      <BccBadge context="mongoose">mongoose</BccBadge>
      <BccBadge context="brand">brand</BccBadge>
    </div>
  `,
});

/**
 * Set the `contrast` prop to control the light / dark contrast of the badge
 */
export const Contrast: StoryFn<typeof BccBadge> = () => ({
  components: { BccBadge },
  template: `
    <div class="flex items-start space-x-2 mb-2">
      <BccBadge context="neutral" contrast="light">Light</BccBadge>
      <BccBadge context="danger" contrast="light">Light</BccBadge>
      <BccBadge context="warning" contrast="light">Light</BccBadge>
      <BccBadge context="success" contrast="light">Light</BccBadge>
      <BccBadge context="info" contrast="light">Light</BccBadge>
      <BccBadge context="muddy-waters" contrast="light">Light</BccBadge>
      <BccBadge context="mongoose" contrast="light">Light</BccBadge>
      <BccBadge context="brand" contrast="light">Light</BccBadge>
    </div>
    
    <div class="flex items-start space-x-2">
      <BccBadge context="neutral" contrast="dark">Dark</BccBadge>
      <BccBadge context="danger" contrast="dark">Dark</BccBadge>
      <BccBadge context="warning" contrast="dark">Dark</BccBadge>
      <BccBadge context="success" contrast="dark">Dark</BccBadge>
      <BccBadge context="info" contrast="dark">Dark</BccBadge>
      <BccBadge context="muddy-waters" contrast="dark">Dark</BccBadge>
      <BccBadge context="mongoose" contrast="dark">Dark</BccBadge>
      <BccBadge context="brand" contrast="dark">Dark</BccBadge>
    </div>
  `,
});

/**
 * Set the `icon` prop to show an icon and the `icon-right` prop to put the icon on the right of the badge
 */
export const WithIcon: StoryFn<typeof BccBadge> = () => ({
  components: { BccBadge },
  setup() {
    return { CheckCircleIcon };
  },
  template: `
    <div class="flex items-start space-x-2">
      <BccBadge context="danger" size="xs" :icon="CheckCircleIcon">
        xs, icon left
      </BccBadge>
      <BccBadge context="danger" size="xs" icon-right :icon="CheckCircleIcon">
        xs, icon right
      </BccBadge>
      <BccBadge context="success" size="sm" :icon="CheckCircleIcon">
        sm, icon left
      </BccBadge>
      <BccBadge context="success" size="sm" icon-right :icon="CheckCircleIcon">
        sm, icon right
      </BccBadge>
    </div>
  `,
});
