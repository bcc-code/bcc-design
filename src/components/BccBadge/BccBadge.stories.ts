import BccBadge from "./BccBadge.vue";
import { CheckCircleIcon } from "@bcc-code/icons-vue";

import type { Meta, StoryFn } from "@storybook/vue3";

export default {
  title: "Components/BccBadge",
  component: BccBadge,
  argTypes: {
    context: {
      description: "Which context color to use for the styling of the badge",
      options: ["neutral", "danger", "warning", "success", "info"],
      control: { type: "radio" },
    },
    size: {
      description: "Size of the badge",
      options: ["xs", "sm"],
      control: { type: "radio" },
    },
    iconRight: {
      description: "Pull icon to right side of the contents",
      control: { type: "boolean" },
    },
    slotDefault: {
      name: "default slot",
      description: "The badge text",
    },
  },
} as Meta<typeof BccBadge>;

const Template: StoryFn<typeof BccBadge> = (args) => ({
  components: { BccBadge },
  setup() {
    return { args, CheckCircleIcon };
  },
  template: `
    <BccBadge v-bind="args" :icon="CheckCircleIcon">
      <template #default>
        {{ args.slotDefault }}
      </template>
    </BccBadge>
  `,
});

export const Example = Template.bind({});
Example.args = {
  context: "neutral",
  size: "xs",
  iconRight: false,
  slotDefault: "Example Badge",
};

export const Context: StoryFn<typeof BccBadge> = () => ({
  components: { BccBadge },
  template: `
    <div class="flex items-start space-x-2">
      <BccBadge context="neutral">neutral</BccBadge>
      <BccBadge context="danger">danger</BccBadge>
      <BccBadge context="warning">warning</BccBadge>
      <BccBadge context="success">success</BccBadge>
      <BccBadge context="info">info</BccBadge>
    </div>
  `,
});

export const WithIcon: StoryFn<typeof BccBadge> = () => ({
  components: { BccBadge },
  setup() {
    return { CheckCircleIcon };
  },
  template: `
    <div class="flex items-start space-x-2">
      <BccBadge context="danger" size="xs" :icon="CheckCircleIcon">
        xs, icon left
      </BccBadge>
      <BccBadge context="danger" size="xs" icon-right :icon="CheckCircleIcon">
        xs, icon right
      </BccBadge>
      <BccBadge context="success" size="sm" :icon="CheckCircleIcon">
        sm, icon left
      </BccBadge>
      <BccBadge context="success" size="sm" icon-right :icon="CheckCircleIcon">
        sm, icon right
      </BccBadge>
    </div>
  `,
});
