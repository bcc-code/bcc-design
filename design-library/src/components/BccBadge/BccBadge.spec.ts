import { describe, expect, it } from "vitest";

import { CheckCircleIcon } from "@bcc-code/icons-vue";
import { mount } from "@vue/test-utils";
import BccBadge from "./BccBadge.vue";

describe("BccBadge", () => {
  it("renders a text from the default slot", () => {
    const wrapper = mount(BccBadge, { slots: { default: "Test Badge" } });
    expect(wrapper.text()).toBe("Test Badge");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders a badge with left icon", () => {
    const wrapper = mount(BccBadge, {
      slots: { default: "Badge" },
      props: { icon: CheckCircleIcon },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders a badge with right icon", () => {
    const wrapper = mount(BccBadge, {
      slots: { default: "Badge" },
      props: { iconRight: CheckCircleIcon },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders a badge with icon but on right", () => {
    const wrapper = mount(BccBadge, {
      slots: { default: "Badge" },
      props: { icon: CheckCircleIcon, iconRight: true },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders a badge with both icons", () => {
    const wrapper = mount(BccBadge, {
      slots: { default: "Badge" },
      props: { icon: CheckCircleIcon, iconRight: CheckCircleIcon },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
