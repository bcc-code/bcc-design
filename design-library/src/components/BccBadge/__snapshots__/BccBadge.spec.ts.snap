// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BccBadge > renders a text from the default slot 1`] = `
"<div class="bcc-badge bcc-badge-neutral light xs">
  <!--v-if--><span class="order-2">Test Badge</span>
</div>"
`;
